File: src/token.ts
type TokenType = string

export interface Token {
  type:     TokenType
  literal:  string
}

export const ILLEGAL = 'ILLEGAL'
export const EOF = 'EOF'

// Identifiers + literals
export const IDENT = 'IDENT' // add, foobar, x, y, ...
export const INT = 'INT' // 1343456

// Operators
export const ASSIGN = '='
export const PLUS   = '+'

// Delimiters
export const COMMA = ','
export const SEMICOLON = ';'
export const LPAREN = '('
export const RPAREN = ')'
export const LBRACE = '{'
export const RBRACE = '}'

// Keywords
export const FUNCTION = 'FUNCTION'
export const LET = 'LET'
export const CONST = 'CONST'
export const IF = 'IF'

const keywords = new Map<string, string>([
  ['wouldyoumind', FUNCTION],
  ['permit', LET],
  ['insist', CONST],
  ['mayhap', IF]
])

export function LookupIdent(ident: string): TokenType {
  if (keywords.has(ident)) {
    return keywords.get(ident)!
  }

  return IDENT
}

File: token/token.go
package token

type TokenType string

type Token struct {
	Type    TokenType
	Literal string
}

const (
	ILLEGAL  = "ILLEGAL"
	EOF      = "EOF"
	IDENT    = "IDENT"
	INT      = "INT"
	ASSIGN   = "="
	PLUS     = "+"
	COMMA    = ","
	LPAREN   = "("
	RPAREN   = ")"
	LBRACE   = "{"
	RBRACE   = "}"
	FUNCTION = "FUNCTION"
	LET      = "LET"
	CONST    = "CONST"
	IF       = "IF"
)

var keywords = map[string]TokenType{
	"wouldyoumind": FUNCTION,
	"permit":       LET,
	"insist":       CONST,
	"mayhap":       IF,
}

func LookupIdent(ident string) TokenType {
	if token, exists := keywords[ident]; exists {
		return token
	}

	return IDENT
}
